### PR TITLE
Implement mutation and validation logics of resources in the admission webhook

### PR DESCRIPTION
**Pull request by derekbit**
The general mutation and validation logics in HTTP API functions are moved or re-implemented in the admission webhook.

[longhorn #3562](https://github.com/longhorn/longhorn/issues/3562)

### CODE REVIEW

File: webhook/resources/node/mutator.go
Comment (shuo-wu) [inline webhook/resources/node/mutator.go:67]: Should Longhorn validate the tags in the validator rather than the mutator? Then is the following `json.Marshal()` required?

Similar below.

File: webhook/resources/engineimage/mutator.go
Comment (shuo-wu) [inline webhook/resources/engineimage/mutator.go:54]: Can we allow users to set the customized labels during this CR creation?

File: webhook/resources/backingimage/validator.go
Comment (shuo-wu) [inline webhook/resources/backingimage/validator.go:98]: Is this redundant?

File: manager/volume.go
Comment (shuo-wu) [inline manager/volume.go:213]: Redundant. This is already moved into the validator.

**Review by shuo-wu — COMMENTED** [review]
Great job!
In general LGTM. Hence I didn't take a review line-by-line. If required, I can do it tomorrow.

File: webhook/resources/backingimage/validator.go
Comment (derekbit) [reply webhook/resources/backingimage/validator.go:98]: Hmm?

Comment (shuo-wu) [reply webhook/resources/backingimage/validator.go:98]: If the backing image CR already exists before the creation, the creation function will directly return an error, right?

File: webhook/resources/engineimage/mutator.go
Comment (shuo-wu) [reply webhook/resources/engineimage/mutator.go:54]: Then will op `replace` overwrite the labels customized by users themselves?

Comment (derekbit) [reply webhook/resources/engineimage/mutator.go:54]: Yes. This will overwrite the labels set by users.
Do you think we need to merge the labels set by `GetEngineImageLabels` and users?
Originally, the labels are from `GetEngineImageLabels`, correct?

File: webhook/resources/node/mutator.go
Comment (shuo-wu) [reply webhook/resources/node/mutator.go:67]: is the main mutation in the function `sort.Strings(tags)`? If YES, I think we can let this mutator take over the sorting only.

File: manager/volume.go
Comment (shuo-wu) [reply manager/volume.go:213]: Oh, sorry. It's not in the validator. We may need to move this to the validator.

Comment (derekbit) [reply manager/volume.go:213]: Ah, thank you. I forgot this part.

File: webhook/resources/backingimage/validator.go
Comment (derekbit) [reply webhook/resources/backingimage/validator.go:98]: Good catch. It's redundant, because the creation function won't be executed if the backing image CR already exists.

File: webhook/resources/engineimage/mutator.go
Comment (shuo-wu) [reply webhook/resources/engineimage/mutator.go:54]: IMO, Longhorn just needs to **add and correct** the labels for the **specific** keys.

File: webhook/resources/node/mutator.go
Comment (derekbit) [reply webhook/resources/node/mutator.go:67]: Yes. Sounds good. Will follow your suggestion.

**Review by shuo-wu — CHANGES_REQUESTED** [review]
I think we need to remove the validations from the volume Creation/Update functions in `manager/volume.go`. Furthermore, as I mentioned in the comments, we need to re-visit and simplify the VolumeManager implementation.

**Comment by derekbit** [comment]
> I think we need to remove the validations from the volume Creation/Update functions in `manager/volume.go`. Furthermore, as I mentioned in the comments, we need to re-visit and simplify the VolumeManager implementation.

Agree. 

One question while I was handling the validation function.
What's the purpose of `CheckEngineImageReadiness`?
How do we get the `currentNodeID` after moving to the webhook?

```
func (m *VolumeManager) verifyDataSourceForVolumeCreation(dataSource longhorn.VolumeDataSource, requestSize int64) (err error) {
...
		if snapName := types.GetSnapshotName(dataSource); snapName != "" {
			if _, err := m.GetSnapshot(snapName, srcVolName); err != nil {
				return err
			}
		}
	}
	return nil
}
```

```
func (m *VolumeManager) GetSnapshot(snapshotName, volumeName string) (*longhorn.Snapshot, error) {
	if volumeName == "" || snapshotName == "" {
		return nil, fmt.Errorf("volume and snapshot name required")
	}
	engine, err := m.GetEngineClient(volumeName)
	if err != nil {
		return nil, err
	}
	snapshot, err := engine.SnapshotGet(snapshotName)
```

```
func (m *VolumeManager) GetEngineClient(volumeName string) (client engineapi.EngineClient, err error) {
	var e *longhorn.Engine
...

	if isReady, err := m.ds.CheckEngineImageReadiness(e.Status.CurrentImage, m.currentNodeID); !isReady {
		if err != nil {
...
```

**Comment by shuo-wu** [comment]
When `volume.spec.DataSource` is set, it indicates that the new volume is a clone of an old volume. In Longhorn, the clone is similar to rebuilding but involves 2 volumes. You can check this for details: https://github.com/longhorn/longhorn/issues/1815

File: manager/node.go
Comment (derekbit) [reply manager/node.go:106]: @shuo-wu 
I will create a ticket for `volumeManager` refactoring, okay?

Comment (shuo-wu) [reply manager/node.go:106]: Sure. But you still need to move the validation/mutation logic to the webhook in this PR......

Comment (derekbit) [reply manager/node.go:106]: Yep.

Comment (derekbit) [reply manager/node.go:106]: @shuo-wu 
Some **update** validation and mutation cannot be moved to the webhook, because there are conflicts between the actions. https://github.com/longhorn/longhorn/issues/3562#issuecomment-1041214241
Thus, I only move the general validation and mutation logics to the webhook.
WDYT?

Comment (shuo-wu) [reply manager/node.go:106]: OK, YES. We just need to move the general checks to the webhook.

File: manager/volume.go
Comment (shuo-wu) [inline manager/volume.go:166]: Can we correct the naming in the mutator?

File: webhook/resources/volume/validator.go
Comment (shuo-wu) [inline webhook/resources/volume/validator.go:113]: Notice that this function has both the old obj and the new obj, we can move more validations from the update APIs to this validator. e.g., once there is a change for field `Spec.NumberOfReplicas`, we need to apply these checks https://github.com/longhorn/longhorn-manager/blob/c04eac1c804597c78d5962c1758e8ed80d2abb37/manager/volume.go#L1016-L1024

And this logic can be applied to other validators. If this would take time, you can open one more PR for it.

File: webhook/resources/volume/validator.go
Comment (shuo-wu) [inline webhook/resources/volume/validator.go:99]: IMO, pkg `manager` is mainly used by the functions in pkg `api`. You can move the required util functions to other pkgs like `datastore` instead.

File: webhook/resources/volume/validator.go
Comment (derekbit) [reply webhook/resources/volume/validator.go:113]: > And this logic can be applied to other validators. If this would take time, you can open one more PR for it.

I will process this part in a separate PR.

File: webhook/resources/volume/validator.go
Comment (shuo-wu) [inline webhook/resources/volume/validator.go:93]: NIT: This can be stored in `volumeValidator`.

**Review by shuo-wu — APPROVED** [review]
In general LGTM

File: webhook/resources/node/validator.go
Comment (innobead) [inline webhook/resources/node/validator.go:48]: I think this should validate again `newNode` instead of `oldNode`, because we want to validate the user request instead of the current settings.

File: webhook/resources/node/validator.go
Comment (innobead) [inline webhook/resources/node/validator.go:59]: Don't these have been covered by OpenAPI schema validation? `+kubebuilder:validation:Minimum`

File: webhook/resources/node/validator.go
Comment (innobead) [inline webhook/resources/node/validator.go:75]: `newNode.Spec.EngineManagerCPURequest` should be always > 0 here, so probably we don't need to have if check here anymore.

File: webhook/resources/node/validator.go
Comment (innobead) [inline webhook/resources/node/validator.go:83]: Don't need this if check, because the same as `newNode.Spec.EngineManagerCPURequest`.

File: webhook/resources/engineimage/validator.go
Comment (innobead) [inline webhook/resources/engineimage/validator.go:40]: Can we remove `// +optional` from `Image`? so we leave the validation to the OpenAPI scheme itself instead.

File: webhook/resources/engineimage/validator.go
Comment (innobead) [inline webhook/resources/engineimage/validator.go:58]: Could we add `refCount` into the message?

File: webhook/resources/recurringjob/validator.go
Comment (innobead) [inline webhook/resources/recurringjob/validator.go:42]: Can we use `+kubebuilder:validation:Enum` to do schema validation instead?

File: webhook/resources/recurringjob/validator.go
Comment (innobead) [inline webhook/resources/recurringjob/validator.go:72]: The same here by using schema validation instead.

File: manager/backingimage.go
Comment (innobead) [inline manager/backingimage.go:75]: Do we still need this? it seems we already cover in the mutator and validator.

File: webhook/resources/backingimage/mutator.go
Comment (innobead) [inline webhook/resources/backingimage/mutator.go:62]: BTW, could we add enum schema validation for `SourceType`?

File: webhook/resources/backingimage/validator.go
Comment (innobead) [inline webhook/resources/backingimage/validator.go:88]: Suggest using `enum` validation instead, so we don't need to add this default error here.

File: webhook/resources/backingimage/validator.go
Comment (innobead) [inline webhook/resources/backingimage/validator.go:99]: Suggest having `cannot delete backing image` as prefix for this error like you have done for others if needed.

**Review by innobead — CHANGES_REQUESTED** [review]

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:59]: Yes, we can remove the checks.

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:48]: Should be `newNode` here.

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:75]: https://github.com/longhorn/longhorn-manager/blob/master/k8s/pkg/apis/longhorn/v1beta2/node.go#L79

Should replace `Minimum=0` with `exclusiveMinimum=0`

Comment (derekbit) [reply webhook/resources/node/validator.go:75]: https://github.com/longhorn/longhorn-manager/blob/master/upgrade/v110to111/upgrade.go#L115
Also need to mutate the `EngineManagerCPURequest` with zero value

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:83]: https://github.com/longhorn/longhorn-manager/blob/master/upgrade/v110to111/upgrade.go#L119

File: manager/backingimage.go
Comment (derekbit) [reply manager/backingimage.go:75]: Yep, should remove this part.

File: webhook/resources/backingimage/mutator.go
Comment (derekbit) [reply webhook/resources/backingimage/mutator.go:62]: Yep, we should add `enum` for these data types.

File: webhook/resources/volume/mutator.go
Comment (innobead) [inline webhook/resources/volume/mutator.go:166]: Can we move this part as part of the same condition above? `volume.Spec.FromBackup != ""`

File: deploy/install/02-components/05-webhook.yaml
Comment (innobead) [inline deploy/install/02-components/05-webhook.yaml:123]: Why we need this?

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, just left some comments needed to resolve before approval. This is a good start to move out some validation logic from the existing code to the same place.

File: webhook/resources/engineimage/validator.go
Comment (derekbit) [reply webhook/resources/engineimage/validator.go:42]: `Image` is required field and non-empty sting.
Remove `// +optional` and Add `// +kubebuilder:validation:MinLength:=1`

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:70]: https://github.com/longhorn/longhorn-manager/blob/555b5a38f57b1875af37614467623010f4d799b8/controller/controller_manager.go

If `EngineManagerCPURequest` or `ReplicaManagerCPURequest` is 0, the default value will be set.
Thus, `0` is also available value here.

File: webhook/resources/engineimage/validator.go
Comment (innobead) [reply webhook/resources/engineimage/validator.go]: yeah, right now it's based on the error template of openAPI schema validation.

**Review by innobead — COMMENTED** [review]
LGTM, but as discussed, need to check if enum validation work as expected first before merge.

**Review by innobead — APPROVED** [review]

File: webhook/resources/volume/mutator.go
Comment (jenting) [inline webhook/resources/volume/mutator.go:208]: Shouldn't we perform the check as lines 80-169 because the user probably changing the Volume CR.
In this case, the mutating webhook of the _Update_ operation would be called, right?

**Review by jenting — COMMENTED** [review]
Sorry for the late review, a few questions.

File: webhook/resources/backingimage/validator.go
Comment (derekbit) [reply webhook/resources/backingimage/validator.go:37]: `backingImage.update` are called in the HTTP API path and the controllers, so I didn't handle it in this phase.
The validation logics in A and B are different, so need to find the common part and move to the webhook

File: webhook/resources/engineimage/mutator.go
Comment (derekbit) [reply webhook/resources/engineimage/mutator.go:36]: Basically, nothing can be moved to `engineImage.Update` or `engineImage.UpdateStatus`..

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:36]: Yep. We probably can implement the `Create`.

File: webhook/resources/volume/mutator.go
Comment (derekbit) [reply webhook/resources/volume/mutator.go:90]: Yep, duplicate.

File: webhook/resources/volume/mutator.go
Comment (derekbit) [reply webhook/resources/volume/mutator.go:208]: Ideally, we should check as line 80-169 in `Update`, but there are some exclusive validations between `Update` operates. As [1], the next step is to find out more common and general checks and move them to the webhook.

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:36]: The `Node.Create` are mainly related to the mutation, so the node validation does not have `Create` function.

File: webhook/resources/engineimage/mutator.go
Comment (jenting) [reply webhook/resources/engineimage/mutator.go:36]: Okay, I was thinking we could mutate the `spec.image`, but this would cause the EngineImage CR name to be changed.

File: webhook/resources/node/validator.go
Comment (jenting) [reply webhook/resources/node/validator.go:36]: If I understand correctly, we don't have to validate the new Node CR such as the `spec.engineManagerCPURequest` and `spec.replicaManagerCPURequest` is valid or not?

Comment (derekbit) [reply webhook/resources/node/validator.go:36]: https://github.com/longhorn/longhorn-manager/blob/master/datastore/longhorn.go#L1767
I'm thinking if we should move this part to the mutation function.
If the node.disks is not set, then set the default disk.
But the `disks` is not a MUST set field.

@jenting, @innobead WDYT?

File: webhook/resources/volume/mutator.go
Comment (jenting) [reply webhook/resources/volume/mutator.go:90]: Should we remove it?

Comment (derekbit) [reply webhook/resources/volume/mutator.go:90]: Yep, will send a PR to remove it.

File: webhook/resources/volume/mutator.go
Comment (jenting) [reply webhook/resources/volume/mutator.go:208]: I see, thank you for the effort.

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:36]: > `spec.replicaManagerCPURequest`

Yes. But some of the values are validated by schema validation.

Comment (jenting) [reply webhook/resources/node/validator.go:36]: Do you mean this part? https://github.com/longhorn/longhorn-manager/blob/3d7e218bc74e1966360bb1aa311b9eee87684efe/datastore/longhorn.go#L1786-L1801

File: webhook/resources/backingimage/validator.go
Comment (jenting) [reply webhook/resources/backingimage/validator.go:37]: I see, thank you for the effort.

File: webhook/resources/node/validator.go
Comment (derekbit) [reply webhook/resources/node/validator.go:36]: > Do you mean this part?
> 
> https://github.com/longhorn/longhorn-manager/blob/3d7e218bc74e1966360bb1aa311b9eee87684efe/datastore/longhorn.go#L1786-L1801

Yep!

Comment (jenting) [reply webhook/resources/node/validator.go:36]: > I'm thinking if we should move this part to the mutation function.
If the node.disks is not set, then set the default disk.

From the comment https://github.com/longhorn/longhorn-manager/blob/3d7e218bc74e1966360bb1aa311b9eee87684efe/datastore/longhorn.go#L1789-L1791
I think we could add to Create mutator only, but not in Update mutator.

Comment (derekbit) [reply webhook/resources/node/validator.go:36]: Just aware of the logic including `disks, err := types.CreateDefaultDisk(dataPath) `. 
Thus, we cannot move it to the admission webhook, or the `dataPath` will be created on a wrong node.
`initDaemonNode`, `CreateDefaultNode` and `CreateDefaultDisk` will be executed while starting the longhorn-manager daemon.

File: webhook/resources/engineimage/mutator.go
Comment (derekbit) [reply webhook/resources/engineimage/mutator.go:36]: Gotcha. Normalizing the `spec.image` and mutating it is good idea, but the previous logic does not include it. Thus, I did't do it to avoid unexpected side effects.

File: webhook/resources/node/validator.go
Comment (jenting) [reply webhook/resources/node/validator.go:36]: I see

> Just aware of the logic including `disks, err := types.CreateDefaultDisk(dataPath) `. Thus, we cannot move it to the admission webhook, or the `dataPath` will be created on a wrong node.
> 
> Currently, `initDaemonNode`, `CreateDefaultNode` and `CreateDefaultDisk` will be executed while starting the longhorn-manager daemon.

I see.

Comment (derekbit) [reply webhook/resources/node/validator.go:36]: > I see
> 
> > Just aware of the logic including `disks, err := types.CreateDefaultDisk(dataPath) `. Thus, we cannot move it to the admission webhook, or the `dataPath` will be created on a wrong node.
> > Currently, `initDaemonNode`, `CreateDefaultNode` and `CreateDefaultDisk` will be executed while starting the longhorn-manager daemon.
> 
> I see.

Leaving a note here.
The admission webhook is responsible for the parameters validation and mutation. Any manipulation of the resources of a node should not be allowed, and the manipulations should handle in the longhorn-manager daemon or controllers.

Comment (innobead) [reply webhook/resources/node/validator.go:36]: @derekbit as discussed, suggest having a note in that package like readme, so it's better to know what kind of code should be put there.

**Pull request merged**
